src/resolve/resolved-dns-query.c

```diff
@@ -524,6 +524,15 @@ int dns_query_new(
 
         assert(m);
 
+        /* Check for records that is refused and refuse query for the records if matched in configuration */
+        DNS_QUESTION_FOREACH(key, question_utf8)
+                if (set_contains(m->refuse_record_types, INT_TO_PTR(key->type)))
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOANO), "Got request for %s record that is refused.", dns_type_to_string(key->type));
+
+        DNS_QUESTION_FOREACH(key, question_idna)
+                if (set_contains(m->refuse_record_types, INT_TO_PTR(key->type)))
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOANO), "Got request for %s record that is refused.", dns_type_to_string(key->type));
```

Comment:
please split out the check into a helper func, i.e.

```c
bool dns_question_has_record_types(DnsQuestion *q, Set *types) { 
        …
}
```

so that these two loops here become:

```c
if (dns_question_has_record_types(question_utf8, m->refuse_record_types) ||
        dns_question_has_record_types(question_idna, m->refuse_record_types))
        return log_debug_errno(…
```

you get the idea

Reply:
hmm, so thining about this: i think this needs some tweaking. DnsQuestion can contain multiple lookup keys primarily to cover for cases where we want to look for A and AAAA for some domain at the same time. Given that this is the primary usecase I think we really should tweak behaviour here so that:

1. when *all* of the request types in the DnsQuestion are among the record types to refuse we return a failure, as your current code does it
2. when however only *some* of the request types in the DnsQuestion are to be refused we should just silently remove them from the question, and continue with the ones that are allowed.

Reply:
Need sometime to digest this. I'll try to reread it again to better understand it.

Reply:
Realised this when i tried to query using resolvectl when trying to build tests for resolvectl. So maybe this is your concern, i'll take a look on how to go about this.

```
$ resolvectl query google.com
google.com: resolve call failed: DNS query type refused.
```